kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/admin/ProcessAdminServiceBase.java

```diff
@@ -101,21 +102,40 @@ public MigrationReportInstanceList migrateProcessInstances(String containerId, L
                                                                                String targetContainerId,
-                                                                               String targetProcessId,
                                                                                String payload, 
                                                                                String marshallingType) {
 
-        ProcessInstanceDesc pi = runtimeDataService.getProcessInstanceById(processInstanceId);
+        ProcessInstanceDesc pi = runtimeDataService.getProcessInstanceById(processInstanceId.longValue());
         if(pi.getParentId() > 0) {
             throw new IllegalArgumentException("Only root process can invoke this migration with all subprocesses");
         }
-        List<Long> processInstancesId =  runtimeDataService.getProcessInstancesWithSubprocessByProcessInstanceId(processInstanceId, 
+
+        MigrationSpecification migrationSpecification = new MigrationSpecification();
+        if (payload != null) {
+            logger.debug("About to unmarshal node mapping from payload: '{}' using container {} marshaller", payload, containerId);
+            migrationSpecification = marshallerHelper.unmarshal(containerId, payload, marshallingType, MigrationSpecification.class);
+        }
+
+        List<Long> processInstancesId =  runtimeDataService.getProcessInstancesWithSubprocessByProcessInstanceId(processInstanceId.longValue(), 
                                                                                                                  singletonList(ProcessInstance.STATE_ACTIVE), 
                                                                                                                  new QueryContext(0, -1)).stream().map(ProcessInstanceDesc::getId).collect(Collectors.toList());
 
-        return migrateProcessInstances(containerId, processInstancesId, targetContainerId, targetProcessId, payload, marshallingType);
+        List<MigrationReport> reports = new ArrayList<>();
+        for(Long processInstanceToMigrateId : processInstancesId) {
+            ProcessInstanceDesc piToMigrate = runtimeDataService.getProcessInstanceById(processInstanceToMigrateId);
+            Optional<MigrationProcessSpecification> spec = migrationSpecification.getProcesses().stream().filter(e -> piToMigrate.getProcessId().equals(e.getSourceProcessId())).findFirst();
+            if(!spec.isPresent()) {
+                logger.error("MigrationProcessSpecification is not correct. Process Instance Id " + processInstanceToMigrateId + " won't be migrated");
+                continue;
+            }
+            String targetProcessId = spec.get().getTargetProcessId();
+            Map<String, String> nodeMapping = spec.get().getNodes();
+            reports.add(processInstanceMigrationService.migrate(containerId, processInstanceToMigrateId, targetContainerId, targetProcessId, nodeMapping));
+        }
+
```

Comment:
Shouldn't we also check for the parent presence, in the `MigrationSpecification` object and throw an exception if it's not coming?

Reply:
nope... we migrate what we can.